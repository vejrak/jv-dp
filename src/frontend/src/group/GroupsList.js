// @flow

import React from 'react'
import { Button } from 'react-bootstrap'
import filter from 'lodash/filter'
import { deleteGroup, editGroup, getGroups } from './actions'
import { canShowUpdateForm } from '../helpers'
import GroupForm from './form/GroupFormConnected'
import type { ErrorResponse, Group } from '../types'
import Spinner from '../components/Spinner'

type Props = $ReadOnly<{|
  deleteGroup: typeof deleteGroup,
  editGroup: typeof editGroup,
  getGroups: typeof getGroups,
  isFetching: boolean,
  error: ?ErrorResponse,
  groups: Array<Group>,
|}>

type State = {
  showEditFormsFor: Array<string>,
}

class GroupsList extends React.PureComponent<Props, State> {
  state = {
    showEditFormsFor: [],
  }

  componentDidMount() {
    this.props.getGroups()
  }

  handleEditClick = (_id: string): void => {
    const { showEditFormsFor } = this.state
    this.setState({ showEditFormsFor: showEditFormsFor.concat(_id) })
  }

  handleHideClick = (_id: string): void => {
    const { showEditFormsFor } = this.state
    this.setState({
      showEditFormsFor: filter(showEditFormsFor, _id),
    })
  }

  updateGroup = (_id: string): Function => (group: Group): void => {
    this.props.editGroup(group, _id)
  }

  handleDeleteClick = async (_id: string) => {
    await this.props.deleteGroup(_id)
  }

  render() {
    const { showEditFormsFor } = this.state
    const { error, isFetching, groups } = this.props

    if (isFetching) return <Spinner />

    if (error)
      return (
        <div className="alert alert-danger" role="alert">
          {error.errmsg}
        </div>
      )

    if (!groups || groups.length === 0) return null
    return (
      <table className="table">
        <thead>
          <tr>
            <th scope="col">Name</th>
            <th />
            <th />
          </tr>
        </thead>
        {groups.map((group) => (
          <tbody key={group._id} className="mb-4">
            <tr>
              <td>{group.name}</td>
              <td>
                <Button
                  className="mt-3 mb-3 btn d-block btn-danger"
                  onClick={() => {
                    if (
                      window.confirm(
                        'Are you sure you wish to delete this item?',
                      )
                    )
                      this.handleDeleteClick(group._id)
                  }}
                >
                  Delete
                </Button>
              </td>
              {!canShowUpdateForm(showEditFormsFor, group._id) && (
                <td>
                  <Button
                    className="mt-3 mb-3"
                    onClick={() => this.handleEditClick(group._id)}
                  >
                    Edit
                  </Button>
                </td>
              )}
            </tr>
            {canShowUpdateForm(showEditFormsFor, group._id) && (
              <>
                <tr>
                  <td colSpan="2">
                    <GroupForm
                      name={group.name}
                      updateGroup={this.updateGroup(group._id)}
                    />
                  </td>
                  <td>
                    <Button
                      className="mt-3 mb-3"
                      onClick={() => this.handleHideClick(group._id)}
                    >
                      Hide
                    </Button>
                  </td>
                </tr>
              </>
            )}
          </tbody>
        ))}
      </table>
    )
  }
}

export default GroupsList
