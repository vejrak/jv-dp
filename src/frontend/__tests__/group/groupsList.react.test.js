import React from 'react'
import renderer from 'react-test-renderer'
import GroupsList from '../../src/group/GroupsList'

const mockProps = {
  deleteGroup: () => {},
  editGroup: () => {},
  getGroups: () => {},
  isFetching: false,
  error: null,
  groups: [{ _id: 'id1', name: 'name1' }],
}

it('renders correctly', () => {
  const tree = renderer.create(<GroupsList {...mockProps} />).toJSON()
  expect(tree).toMatchSnapshot()
})
