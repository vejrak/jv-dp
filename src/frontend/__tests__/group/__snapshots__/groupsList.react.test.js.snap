// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<table
  className="table"
>
  <thead>
    <tr>
      <th
        scope="col"
      >
        Name
      </th>
      <th />
      <th />
    </tr>
  </thead>
  <tbody
    className="mb-4"
  >
    <tr>
      <td>
        name1
      </td>
      <td>
        <button
          className="mt-3 mb-3 btn d-block btn-danger btn btn-primary"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Delete
        </button>
      </td>
      <td>
        <button
          className="mt-3 mb-3 btn btn-primary"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Edit
        </button>
      </td>
    </tr>
  </tbody>
</table>
`;
